pytorch_finufft/functional.py

```diff
@@ -72,108 +72,191 @@ def coordinate_ramps(shape, device):
+
+        if batch_points is not None:
+            # need a for-loop here
+            points = points.movedim(batch_points, 0)
+            if batch_values is not None:
+                output = torch.stack(
+                    [
+                        FinufftType1.apply(
+                            points[i],
+                            values[i],
+                            output_shape,
+                            finufftkwargs,
+                        )
+                        for i in range(info.batch_size)
+                    ],
+                    dim=0,
+                )
+            else:
+                output = torch.stack(
+                    [
+                        FinufftType1.apply(
+                            points[i],
+                            values,
+                            output_shape,
+                            finufftkwargs,
+                        )
+                        for i in range(info.batch_size)
+                    ],
+                    dim=0,
+                )
+        else:
+            output = FinufftType1.apply(points, values, output_shape, finufftkwargs)
+
+        return output, 0
+
     @staticmethod
     def backward(  # type: ignore[override]
         ctx: Any, grad_output: torch.Tensor
     ) -> Tuple[Union[torch.Tensor, None], ...]:
-        """
-         Implements derivatives wrt. each argument in the forward method.
-
-         Parameters
-         ----------
-         ctx : Any
-             Pytorch context object.
-         grad_output : torch.Tensor
-             Backpass gradient output
-
-         Returns
-         -------
-        Tuple[Union[torch.Tensor, None], ...]
-             A tuple of derivatives wrt. each argument in the forward method
-        """
         _i_sign = -1 * ctx.isign
         _mode_ordering = ctx.mode_ordering
         finufftkwargs = ctx.finufftkwargs
 
         points, values = ctx.saved_tensors
+        points = torch.atleast_2d(points)
+
         device = points.device
+        ndim = points.shape[0]
 
         grads_points = None
         grad_values = None
 
-        ndim = points.shape[0]
-
         nufft_func = get_nufft_func(ndim, 2, device.type)
 
-        if any(ctx.needs_input_grad) and _mode_ordering:
-            grad_output = torch.fft.fftshift(grad_output)
+        if any(ctx.needs_input_grad):
+            if _mode_ordering:
+                grad_output = batch_fftshift(grad_output, ndim)
+
+            # group together batched dimensions, if any
+            shape = grad_output.shape[-ndim:]
+            batch_dims = grad_output.shape[:-ndim]
+            batched_grad_output = grad_output.reshape(-1, 1, *shape)
+            nbatch = batched_grad_output.shape[0]
 
         if ctx.needs_input_grad[0]:
             # wrt points
-            coord_ramps = coordinate_ramps(grad_output.shape, device)
+            coord_ramps = coordinate_ramps(shape, device)
+
+            # nbatch x ndims x ...
```

Comment:
do not understand this wrt the line below it. ndims == 1?

Reply:
the dim 1 created in the line below is the one that the coord ramp directions will live in as far as I understand. So it will be ndims after that product has been computed